File: Assets/Resources/Scripts/TerrainSpawner.js
var spawnee: Transform;
var spawnee2: Transform;
var spawnee3: Transform;
var spawnees : Transform[];
private var newgrounds : Transform;
private var spawn_array = new Array();
private var start_ground_amount = 3;
private var ground_length = 20;
private var player_start;
private var ground_start;
private var spawn_prefab : Transform;

private var playa ;
private var ground_count = 0;
function Start(){
	playa = GameObject.Find("Player");
	player_start = playa.transform.position.x;
	ground_start = player_start;
	for(ground_count=0;ground_count< start_ground_amount; ground_count++){
		newgrounds = Instantiate(spawnees[0],Vector3(player_start+(ground_count-1)*ground_length,0,0),Quaternion.identity);
		spawn_array.push(newgrounds);
	}

	
}

function Update () {
	var dmw;
	/*
	if (Input.GetKeyDown("up")){
		newgrounds = Instantiate(spawnees[0],Vector3(playa.transform.position.x,playa.transform.position.y - 6,0),Quaternion.identity);
		spawn_array.push(newgrounds);
	}
	if (Input.GetKeyDown("down")){
		if (spawn_array.length > 0){
			dmw = spawn_array.shift();
			Destroy(dmw.gameObject);
		}	
	}*/
	
	if(playa.transform.position.x>player_start+ground_length)
	{
		var segment_num = Mathf.Floor(Random.value*3); //genrate random number from 0-2
		var segment_length;
		var rotation = Quaternion.identity;
		rotation.eulerAngles = Vector3(0,0,20);
		switch (segment_num)
		{
			case 0:
				//spawn_prefab = Resources.Load("Terrain01") as Transform;
				newgrounds = Instantiate(spawnees[0],Vector3(ground_start+((ground_count++-1)*ground_length),0,0),Quaternion.identity);				
				break;
			case 1:
				//spawn_prefab = Resources.Load("Terrain02")as Transform;
				newgrounds = Instantiate(spawnees[1],Vector3(ground_start+((ground_count++-1)*ground_length),3.92544,0),rotation);
				break;
			case 2:
				//spawn_prefab = Resources.Load("Terrain03") as Transform;
				newgrounds = Instantiate(spawnees[2],Vector3(ground_start+((ground_count++-1)*ground_length-4),8.5,0),Quaternion.identity);				
				break;
			default:
				break;
		}
		//newgrounds = Instantiate(spawnee,Vector3((ground_count++-1)*ground_length,0,0),Quaternion.identity);
		spawn_array.push(newgrounds);
		
		if (spawn_array.length > 0){
			dmw = spawn_array.shift();
			Destroy(dmw.gameObject);
		}	
		player_start += ground_length;
	}
	
}

File: Assets/Standard Assets/CustomScripts/TerrainSpawner.cs
using UnityEngine;
using System.Collections;
using System.Collections.Generic;



public class TerrainSpawner : MonoBehaviour {
	
	//Arrays that describe the various terrain pieces//////////////////
	public GameObject[] spawnees;
	public float[] spawneeLengths;
	public float[] spawneeXOffsets;
	public float[] spawneeYOffsets;
	public float[] spawneeAngles; 
	private Quaternion[] spawneeRotations; //used privately  to convert angles to rotations

	private GameObject newGrounds;
	private Queue<GameObject> spawnedTerrain = new Queue<GameObject>(); //holds references to spawned objects
	private int startGroundAmount = 3;
	private float groundLength = 20;
	private float spawnStart = 0;
	private float groundStart = 0;
	private Transform spawnPrefab;
	
	
	private GameObject playa ; //reference to player
	private int groundCount = 0; //iterator for ground loops
	private float pastXOffset =0; 
	private float lastSpawnPos = 0;
	// Use this for initialization
	void Start () {
		playa = GameObject.Find("Player");// get reference to player
		spawnStart = playa.transform.position.x; //initiate start of spawns to initial player position + half of first terrain length
		groundStart = spawnStart; //start of ground is first spawn point
		
		///////////Generate intial platoforms remeber to randomize later /////////////////////////
		for(;groundCount< startGroundAmount; groundCount++){
			Vector3 startVector = new Vector3(spawnStart+(groundCount-(float)0.5)*spawneeLengths[0],0,0);
			newGrounds = Instantiate(spawnees[0],startVector,Quaternion.identity) as GameObject;
			spawnedTerrain.Enqueue(newGrounds);
		}
		//setup variables needed to calculate next platform position
		lastSpawnPos = spawnStart + spawneeLengths[0]*(float)1.5; //set last spawnpoint to position of last spawned terrain
		pastXOffset = spawneeLengths[0]*(float)0.5 - spawneeXOffsets[0]; // set previous  x offset for next psawn calculation
		spawnStart += spawneeLengths[0]; //set next spawnpoint
		
		/////////////////////////////////////////////////////////////////
		
		//translate inspector inputted angles into actual Rotations/////////////////
		spawneeRotations = new Quaternion[spawnees.Length];
		for(int i=0; i<spawnees.Length; i++)
		{
			spawneeRotations[i].eulerAngles = new Vector3(0,0,spawneeAngles[i]); 
		}
		/////////////////////////////////////////////////////////////////////////////
	
	}
	
	// Update is called once per frame
	void Update () {

		
		GameObject obsoleteTerrain;
		if(playa.transform.position.x>spawnStart) // when player goes past point where another land needs to be spawned
		{

			int segmentNum = (int)Mathf.Floor(Random.value*spawnees.Length); //genrate random number from 0- number of terrain types - 1
			
			//spawn randomly chosen piece  with offset due differing centerpoints
			float newSpawnPos = lastSpawnPos + pastXOffset + (spawneeLengths[segmentNum]/2) + spawneeXOffsets[segmentNum];
			newGrounds = Instantiate(spawnees[segmentNum],new Vector3(newSpawnPos,spawneeYOffsets[segmentNum],0),spawneeRotations[segmentNum])as GameObject;				
		
			
			spawnedTerrain.Enqueue(newGrounds); //store reference to new land
			spawnStart += spawneeLengths[segmentNum];	//set new spawn point
			pastXOffset = (spawneeLengths[segmentNum]/2)-spawneeXOffsets[segmentNum]; //set up the previous part of the next offsets
			lastSpawnPos = newSpawnPos; // record current spawn position for next spawn 
			if (spawnedTerrain.Count > 0){ // delete old terrain if it exists
				//print(spawnedTerrain.Dequeue());
				obsoleteTerrain = spawnedTerrain.Dequeue();
				
				Destroy(obsoleteTerrain);
			}	
			
		}
	}
}
